### DISCUSSION TITLE
Support built-in emoji selections

### DISCUSSION THREAD

**Discussion by MaethorNaur** [8VIM/8VIM#438] [Polls]
Currently, emojis rely on switching keyboards.
In most cases this is not needed to switch keyboard has an emoji selections.
But this is sometimes not possible, forcing us to switch to that keyboard and back to 8vim manually.

Should 8Vim have a built-in emoji selection without switching the keyboard?

refs: #435

**Comment by alterNERDtive** [comment]
Over there you said

> This is doable but the current idea is to rely on the app emoji selection (**most apps have one**)

(emphasis mine)

I disagree. Quite heavily. Emojis have gone far beyond what used to be “smileys” back in the day. That means there are way more use cases for them than there used to be, too.

At the same time, it’s all Unicode. It’s text input. So any application that takes text input can generally take emoji input as well. I’d wager that the _vast_ majority of applications that _can_ take emojis does _not_ have built-in support for them.

On the desktop I’ve grown so accustomed to using emojis that I have the most commonly used ones in my compose file, and use https://github.com/biox/shmoji + dmenu to have access to the full list _anywhere_. On the phone they are still a royal PITA.

So, in case that wasn’t obvious, I voted “Yes”. As to what an actual implementation / UI would look like, no idea yet.

**Reply by MaethorNaur** [reply]
The idea was to have a dedicated panel with all emojis which will be inserted into the input without having to switch back and forth between keyboards in case of the app doesn't have a dedicated panel for it

**Reply by alterNERDtive** [reply]
Yeah, I don’t think trying to integrate them into the “main” input circle would work. So, like the numpad thing, it would definitely be a dedicated panel.

Search would be great (actually I don’t think it’ll be feasible w/o search unless you limit the thing to a very small subset), and I really like the idea of suggestions while typing. Though actually (without my phone rn) I’m not sure if 8Vim even has suggestions. Either it doesn’t or I have them turned off ;)

**Comment by cmcaine** [comment]
I'd appreciate this because:

1. emoji and other non-verbal communication is important https://www.multiamory.com/podcast/335-nonverbal-communication
2. Some of the apps I use don't have emoji selection widgets (e.g. text input in browsers) and I would anyway prefer to have a consistent widget in the keyboard rather than a different one for each app
3. switching to the emoji keyboard doesn't work for me: it changes the active keyboard, the default keyboard and also doesn't open the emoji panel on the new keyboard.

**Comment by mtivi** [comment]
The lack of emojis is the most annoying thing about this keyboard. The extra flow of _switching to a different keyboard_ often means I'll switch keyboards for a couple days. 

It seems like a _weird_ decision to force people into an alternative keyboard just to type an emoji. Emojis have such widespread usage that there is a system emoji picker for [Windows](https://support.microsoft.com/en-us/windows/588e0b72-0fff-6d3f-aeee-6e5116097942),  [Mac](https://osxdaily.com/2015/05/27/quick-type-emoji-mac-keyboard-shortcut/), [Gnome](https://www.omgubuntu.co.uk/2018/06/use-emoji-linux-ubuntu-apps), _and_ [Chrome OS](https://beebom.com/how-use-emojis-chromebook/).

[Emojipedia has the full list of Emojis for Android](https://emojipedia.org/google)l Unicode.com has a[ list of all emojis, as well as a text description ](https://unicode.org/emoji/charts/full-emoji-list.html) (which could be used for searching).

In terms of implementation, I _personally_ really like gboard's emoji keyboard. A search bar, little icons for categories, and a few other integrations. There's also other integrations, such as emoticons, gifs, stickers for people that use them. I also miss the feature where I can type a word, and autocomplete will give me the option to change the word I just typed into an emoji. It's practical, because there's _so many _emojis,_ that usually I'd have to use the search bar anyway. 

Anyway, I really hope this feature is implemented. It's been a personal pain point for quite some time now.

**Comment by ManDay** [comment]
I agree that Emoji input should be readily accessible from any keyboard. Whether "built-in" or through a different program ("Intent", in Android lingo) seems to be irrelevant, as long as the flow/UX works sufficiently well.

As for the "urgency" of the matter, I agree that the current method of simply dropping into another keyboard is not sufficiently fluent and clumsy to the point where it becomes unfeasible. For people like me, who have the sidebar disabled, it is outright impossible.

OTOH, most messaging apps provide an in-app Emoji picker, which kind of mitigates the problem to a certain point. Then again, there are apps without such a picker (which is conceptually the right thing to do), where entering as much as a single Emoji becomes a huge PITA. So yes, a solution is needed.

I think a first would be to provide a **Gesture** for a call to [`showInputMethodPicker()`](https://developer.android.com/reference/android/view/inputmethod/InputMethodManager#showInputMethodPicker()), as by https://github.com/8VIM/8VIM/issues/437#issuecomment-1939108355, that would mitigate both, the lack of Emoji input *and* the lack of integrated layout switching for people without the sidebar.

**Reply by alterNERDtive** [reply]
> As for the "urgency" of the matter, I agree that the current method of simply dropping into another keyboard is not sufficiently fluent and clumsy to the point where it becomes unfeasible.

:+1: simply because I refuse to use GBoard, and my alternate keyboard doesn’t really support emojis either ;)

**Comment by flide** [comment]
Actually, 8vim used to have emojis as part of the keyboard then i split it
into a seperate keyboard. Idea was to allow ppl to use keyboard which are
dedicated for emoji, kanji, gifs, sticker etc. I saw so many of such
keyboards that i assumed it will become  norm.
My future predicting capabilities suck...

**Reply by ManDay** [reply]
In my opinion you're just ahead of your time @flide . If input methods in Android were properly designed, they would indeed be modularized (not only by [input type](https://developer.android.com/reference/android/text/InputType), such as "text", "math", "emoji", which is currently passed in the intent to the keyboard itself; but also by other criteria such as locale or even app) and quickly accessible from eachother.

It's not your fault that this is so poorly executed in AOSP and no seamless way to change between a exhaustive set of input types, but only a clumsy "input method picker" dialog which can be opened from the app. Ideally, it would be exactly as you say!

**Reply by alterNERDtive** [reply]
> My future predicting capabilities suck...

Predictions are hard. Especially if they concern the future.

**Reply by MaethorNaur** [reply]
We could have a compromise. leaving the choice of an external keyboard for emoji and/or an emoji picker, leaving the choice to the user?

**Reply by Gorkycreator** [reply]
I actually really like the idea of a split keyboard and think it is a really flexible/valuable component.

The problem, though, is mostly due to documentation/branding of the emoji board and the lack of attention it has recieved over the years.

There's no mention of the [pictograph keyboard](https://github.com/8VIM/Pictograph-Keyboard) in any onboarding instructions for 8vim. I had to spend a lot of time on google to find it--my first inclanation was not to browse the github page for another keyboard to install... I also did not know that pictograph was associated with 8vim until very recently. That's a bit of a problem... These apps should very clearly point at each other and be recommended to users as the default options.

Once I found it, I liked how pictograph keyboard worked with 8vim! It was very good and using it with 8vim felt like a complete keyboard experience. It was a bit clunky without search, but the main reason I retired it was because it just hasn't been updated to the latest emoji standards. I kept going to it for an emoji, not finding the one I was looking for (like the melting face), and falling back on another method to get it.

With some love and better visibility, I think this keyboard could easily fill most user needs without integrating emoji directly into 8vim.

Functionally, I wouldn't complain about having an emoji panel in 8vim, but I appreciate the niche use case pictograph fills and would love to see an effort to onboard 8vim users before giving up on it. I'd be happy to submit a PR with Readme updates, if that would be valuable.

**Reply by NK308** [reply]
@flide Btw what are those keyboards, which allow someone to insert stickers and GIFs?
How would that even be possible for a system keyboard, since it has nothing to do with inserting unicode into a text field?
I only know of apps, which have such functionality built in, but usually just means browsing an image collection provided by the server of that specific messenger app, and creating either a special non-text message with a reference to that resource, or just sending an image file after downloading it. Both not standardized between apps.

**Reply by flide** [reply]
that is true, gifs or stickers will only work for apps that will accept images within the text box, but i remember there were straight forward ways to test it a text box will accept media or not.

**Comment by flide** [comment]
that would be the way to go. Can't force my choice when I know I am wrong.

**Reply by MaethorNaur** [reply]
You are not wrong, just a different point of view. So I think giving the choice to the user on how to select emoji ("built-in"/external or both) is the best approach.

**Comment by jm355** [comment]
you may be able to use the code for the emoji picker from thumbkey https://github.com/dessalines/thumb-key
